### PR TITLE
Migrate HTTP server to net/http

### LINKED ISSUE #4979: Migrate HTTP server from fasthttp to net/http
Since at least 2020, Dapr has been using the [fasthttp](https://github.com/valyala/fasthttp) library for the internal HTTP API server. An [Architectural Decision Record](https://github.com/dapr/dapr/blob/1cbf67f8a5da0bb5be4e3a123632c17ebb173f4e/docs/decision_records/architecture/ARC-004-http-server.md) from June 2020 (that is still in "proposed" status) contains information on this decision, which was based on the fact that, at least on synthetic benchmarks, fasthttp is optimized for performance and raw speed, primarily thanks to avoiding almost any memory allocation.

However, with more experience building Dapr and with the desire to push it forward, we are now uncovering a series of issues that could make it advisable to convert the internal HTTP server to be based on the Go standard library (net/http). It's likely that our use cases are not what fasthttp is meant for, and as we invest in things like [streaming supports](https://github.com/dapr/dapr/pull/4903), the issues will only be growing. This is advice that was provided by the maintainer (it appears to be a single person now) of fasthttp too.

Although switching from net/http to fasthttp will have a negative performance impact, arguably this should be negligible in the context of Dapr. Each one of our handlers performs a lot of operations (and are in large part I/O bound) and a lot of allocations and spanning of goroutines, which mask the effect of any micro-optimization performed by fasthttp.

> Note that fasthttp contains both a server and client, and Dapr uses both, although the client part plays a very small role.

I am categorizing the issues in 3 groups.

### Lack of context support 

With fasthttp, each request does not get a context that is tied to the lifetime of the request.

Although each handler is passed a `ctx` which does implement the `context.Context` interface (and we use it a lot in Dapr, [example](https://github.com/dapr/dapr/blob/master/pkg/http/api.go#L2181)), the lifecycle of that context is tied to the server's context and not the request's context. This seems to be an [explicit design decision](https://github.com/valyala/fasthttp/issues/1350) (although not clearly documented).

This is an issue for Dapr because if the user (app) cancels a request, we do not have a way to be notified of that and abort in-progress operations.

According to the [fasthttp maintainer](https://github.com/valyala/fasthttp/issues/1219#issuecomment-1041548933):

> If people want to use contexts they really shouldn't be using fasthttp. When you are using contexts it shows that you don't need the performance increase fasthttp gives. If you really needed that performance you wouldn't be using contexts to begin with.

### No support for HTTP/2

fasthttp [does not include](https://github.com/valyala/fasthttp/issues/144) support for HTTP/2. This means application cannot use HTTP/2 to connect to Dapr, possibly missing out on the performance benefits of things like multiplexing, etc.

## Problems with handling streams

At a high level, fasthttp has incomplete support for streaming.

For example, the fasthttp client does not support making requests and sending/reading streams. In the PR adding support for streams, I had no choice but to use net/http for outbound requests (in the pkg/channel/http package, for Dapr to invoke the app).

[According to the maintainer](https://github.com/valyala/fasthttp/issues/1340), streaming support has never been a priority because the kind of apps fasthttp is designed for do not need streaming, as they work with small blobs of data:

> fasthttp is all about performance and it achieves this by not supporting these kinds of things. Streaming of response bodies doesn't make sense in the use cases fasthttp is for. If you need body streaming it's better to use net/http.

More specifically, even with the limited support for streaming, a lot of things are left to the developers (and are generally un-documented):

### No explicit support for streaming response body from the server

In a fasthttp server, responding with a stream is not something that is available as built-in, and requires "hijacking" the request. Example:

https://github.com/dapr/dapr/blob/bcb202c79d41e311b259af5414c355c2be7f1e4d/pkg/http/responses.go#L109-L138

[(Issue opened in fasthttp's repo)](https://github.com/valyala/fasthttp/issues/1343)

### Support for reading requests as a stream seems not mature

The only area where fasthttp does have "complete" support for streams seems to be in reading the request body in the server. However, that was a contribution from the community and it seems implemented in a spotty way. At a high level:

- Even when streaming is enabled, some requests do not come in as streams, so trying to read them with the methods to get a stream returns no data.
- When the request body is coming in as stream, the "max request body size" is ignored

Examples of "workarounds" required:

https://github.com/dapr/dapr/blob/bcb202c79d41e311b259af5414c355c2be7f1e4d/pkg/http/api.go#L2478-L2487

Even more, support for using streams seems not fully stable. In the PR adding streaming support, E2E tests seems more flaky with occasional "EOF" errors. (It appears that it is not just us: [1](https://github.com/valyala/fasthttp/issues/1195) [2](https://github.com/valyala/fasthttp/issues/1353) [3](https://github.com/valyala/fasthttp/issues/1006))

## General consideration

Lastly, a few general considerations:

- As the maintainers made it very clear (and are not shy of repeating to everyone in the repo), fasthttp is optimized for certain use cases that are not really aligned with what we are doing. It seems to shine in situations where it's dealing with small amounts of data and the response is generated synchronously (perhaps apps that closely resemble "helloworld-style benchmarks"). This is different from Dapr, where each handler can take a lot of time processing the request, makes a lot of allocations, and is generally I/O-bound
- The code of fasthttp is very complex and documentation is not always clear (see how we've been using the contexts "wrong" all this time). This is especially true if you are looking at the code that implements the (partial) support for streams.
- Lastly, from a risk perspective, net/http is maintained by the Go team and receives lots of attention. It is tested heavily in production in a variety of scenarios, and it is the target of extensive scrutiny also for security vulnerabilities. At this time, it appears that fasthttp has a single, part-time maintainer. While not uncommon in the OSS space, and not necessarily a deal-breaker per se, given the extent to which Dapr depends on this library, and the fact that servers are often a high-risk point for attacks, it's a point worth considering too in the risk calculations.

### PR DESCRIPTION
**Pull request by ItalyPaleAle**
- Fixes #4979 (although more work will be needed to completely remove fasthttp, see below)
- Necessary (but not quite sufficient) to fix also:
  - #5392
  - #6246
  - #6364
  - microsoft/azure-container-apps#760

This PR changes the HTTP server's listeners from fasthttp to net/http and fixes (or enables fixing) a lot of the scenarios that are blocked by #4979. Immediately, this should enable the web server to use HTTP/2 when TLS is enabled.

## Included and not included

What this PR includes:

- Changes the listeners for the 3 HTTP servers (app, healthz, pprof) from using fasthttp to using the standard library (net/http)
- Changes all built-in middlewares used by the servers above to be based on net/http

What this PR does NOT include:

- No handler has been updated from fasthttp to net/http, and handlers continue to work thanks to the "adapters"
  - This is not something new as in Dapr 1.10, all requests go through adapters twice to be converted fasthttp(listener + built-in middlewares)->net/http (middleware components)->fasthttp (handlers). Now there's only one conversion: net/http(everything but handlers)->fasthttp(handlers)
- Does not change the router, which is still the fasthttp one. This is blocking updating handlers from fasthttp to net/http, and needs to be done in a separate PR.
- Does not enable H2C. Should be done in a separate PR. 
- Some unit tests, including many (not all) in `pkg/http` still use fasthttp to create the test server. This isn't a concern at least until the handlers are migrated, but it should be changed in the future.

All features continue to work as expected, as testified by the fact that tests have not changed. Only one exception to this rule is that when performing HTTP->gRPC invocations, the "dapr-host" metadata is not included anymore (this is probably ok also because HTTP->gRPC service invocation is being deprecated)

## Next steps

- [ ] Update the router to one based on net/http, and apply the adapters only on individual handlers rather than on the router itself.
- [ ] Migrate the Universal API adapter from fasthttp to net/http
- [ ] Migrate APIs that can use streaming (such as service invocation) to be based on net/http and leverage streaming for input and output
- [ ] Migrate all other APIs from fasthttp to either net/http handlers or Universal APIs.

### CODE REVIEW

File: pkg/http/api.go
Comment (ItalyPaleAle) [inline pkg/http/api.go:1393]: Reviewers: this was removed for now since it's not effective (see #6246 )

File: pkg/http/server.go
Comment (ItalyPaleAle) [inline pkg/http/server.go:289]: Reviewers: Dapr is likely behaving incorrectly here. Setting "*" in CORS should mean "allow all origins", but Dapr <=1.10 was actually omitting the CORS headers entirely. Since this is a backwards-incompatible change, I am not touching it now.

File: utils/responsewriter/README.md
Comment (daixiang0) [inline utils/responsewriter/README.md:3]: Why can not import directly? Do we change some codes? If so, please comment changes.

Comment (ItalyPaleAle) [reply utils/responsewriter/README.md:3]: Yes there are significant changes, such as support for user values (which we need when moving from fasthttp - may be removed in the future) and removal of "features" (that negroni only needs): https://github.com/dapr/dapr/pull/6248/files#diff-83f49a9138e9ad18fbee03f55f1035fa5ca1a626e1777ec037752212db0e11b7R74

Also, I would prefer not import the entire negroni framework if we just need a single struct. Importing the entire package would cause a lot more code to be brought in that we don't need.

Comment (daixiang0) [reply utils/responsewriter/README.md:3]: Got, please comment changes in the README.

Comment (ItalyPaleAle) [reply utils/responsewriter/README.md]: Done

**Comment by yaron2** [comment]
> /ok-to-perf

Long have we waited for this change... Very curious about the perf results

**Comment by ItalyPaleAle** [comment]
@yaron2 all perf tests are passing, and looks like perf may actually be **improving**

Looking at the service invocation perf test (which is particularly useful here as it is among those more sensitive to the web server's own perf):

[This PR](https://github.com/dapr/dapr/actions/runs/4722655475)
- P50: 1.38ms
- P90: 1.98ms

Three recent perf tests from master ([1]https://github.com/dapr/dapr/actions/runs/4722541970) [2](https://github.com/dapr/dapr/actions/runs/4686319469) [3](https://github.com/dapr/dapr/actions/runs/4690592444))
- Added latency avg: 1.76ms / 1.50ms / 1.47ms
- P90: 2.86ms / 2.13ms /2.09ms

**Comment by yaron2** [comment]
This is a horrible regression in general, at the very least for 1.9 we had ~1.10ms for P90 and I am quite certain for 1.10 as well (will check later for 1.10.0). The regression in master started before this PR

**Comment by ItalyPaleAle** [comment]
@yaron2 Not sure about 1.9 because logs have expired for those runs.

This is what I could find from the release-1.10 branch:

[First](https://github.com/dapr/dapr/actions/runs/4235229688/jobs/7358708370)
image

[Second](https://github.com/dapr/dapr/actions/runs/4318079447/jobs/7536068522)
image

**Comment by yaron2** [comment]
> @yaron2 Not sure about 1.9 because logs have expired for those runs.
> 
> This is what I could find from the release-1.10 branch:
> 
> [First](https://github.com/dapr/dapr/actions/runs/4235229688/jobs/7358708370) image
> 
> [Second](https://github.com/dapr/dapr/actions/runs/4318079447/jobs/7536068522) image

Right, the first run is an outlier, the 2nd was the more common perf numbers we've seen. Now we need to check what caused the regression

**Comment by ItalyPaleAle** [comment]
In any case, this should not block this PR, correct? As it seems at the very least this isn't causing further regression (and may be having a positive impact actually)

PS: I wonder if it's also possible the perf regression may not be our code but some infra-related thing. Since between the two runs linked above on release-1.10, the biggest difference seems to be time. Just wondering.

**Comment by yaron2** [comment]
We will want to merge this at the beginning of the 1.12 milestone to make sure we can release 1.11 compatible versions to test thoroughly.

**Comment by ItalyPaleAle** [comment]
This is disappointing as we had this scheduled for 1.11. We have work that depends on this right now that was scheduled for 1.11: complete streaming (still in preview), high-level crypto, among others.

**Comment by ItalyPaleAle** [comment]
This is sad as we had this scheduled for 1.11. We have work that depends on this right now that was scheduled for 1.11: complete streaming (still in preview), high-level crypto, among others.

Any chance we can get this into 1.11 and unblock these scenarios?

**Comment by daixiang0** [comment]
Good things are never too late. I agree that merge it later then we have enough time to ensure it is good enough.

Not only for this PR, for other big PRs, I do not want to merge them near the end of the milestone.

**Comment by daixiang0** [comment]
BTW, do we also need to migrate in all components? It looks a big job.

**Comment by yaron2** [comment]
> BTW, do we also need to migrate in all components? It looks a big job.

Middleware components support the net/http signatures

**Comment by ItalyPaleAle** [comment]
@artursouza 

Fixed all the issues reported by the SDK tests:

- Java: issue was with the shutdown sequence. Stopping a net/http server returns an error always, which normally is ErrServerClosed, and had to be ignored.
- Python: issue was related to trace headers not being converted between net/http and fasthttp. The routes are still implemented as fasthttp, but the middleware that sets the tracing span now uses net/http. I added a temporary compatibility layer to make it work (until the routes are converted to net/http)

**Review by yaron2 — APPROVED** [review]

**Pull request merged**
